### PR TITLE
More simplification using static bounds of Params.

### PR DESCRIPTION
**Pull request by jingpu**
This PR adds the ability to simplify bounds using the static bounds of Params.

const_int_bounds() also digs into Uint and Cast to a limited extent.

### CODE REVIEW

**Comment by jingpu** [comment]
It looks like the assertions checking Params will be removed by the simplification pass because with this PR the simplification assumes Params are in their static bounds.
Any suggestion how to work around this issue?

File: src/Simplify.cpp [hunk 1]
```diff
@@ -308,12 +309,28 @@ class Simplify : public IRMutator {
         if (const int64_t *i = as_const_int(e)) {
             *min_val = *max_val = *i;
             return true;
+        } else if (t.is_uint() && is_const(e) && t.is_scalar()
```
Comment (dsharletg) [inline src/Simplify.cpp]: I think this should be implemented in `as_const_int`. It's technically correct (as long as the uint isn't greater than the max of a signed int64), but perhaps confusing to some code that uses `as_const_int`?

File: src/Simplify.cpp [hunk 2]
```diff
@@ -308,12 +309,28 @@ class Simplify : public IRMutator {
...
                 return true;
+            } else if (v->param.defined() && !v->param.is_buffer()) {
+                int64_t min_a, min_b, max_a, max_b;
+                if (v->param.get_min_value().defined() &&
+                    const_int_bounds(v->param.get_min_value(), &min_a, &max_a) &&
+                    v->param.get_max_value().defined() &&
+                    const_int_bounds(v->param.get_max_value(), &min_b, &max_b)) {
+                    *min_val = std::min(min_a, min_b);
+                    *max_val = std::max(max_a, max_b);
+                    bounds_info.push(v->name, { *min_val, *max_val });
```
Comment (dsharletg) [inline src/Simplify.cpp]: Should this be pushing here? None of the other handlers do.

File: src/Simplify.cpp [hunk 3]
```diff
@@ -308,12 +309,27 @@ class Simplify : public IRMutator {
         if (const int64_t *i = as_const_int(e)) {
             *min_val = *max_val = *i;
             return true;
+        } else if (const uint64_t *u = as_const_uint(e)) {
+            if (t.bits() < 64) {
```
Comment (psuriana) [inline src/Simplify.cpp]: It's probably better to use Int(64).can_represent(*u). It also does the overflow check for you.

File: src/Simplify.cpp [hunk 4]
```diff
@@ -436,6 +452,17 @@ class Simplify : public IRMutator {
                 return no_overflow_scalar_int(t.element_of()) ||
                        (t.can_represent(*min_val) && t.can_represent(*max_val));
             }
+        } else if (const Cast *cast = e.as<Cast>()) {
+            int64_t min, max;
+            if ((cast->type.is_int() || cast->type.is_uint()) &&
+                const_int_bounds(cast->value, &min, &max)) {
+                    if (can_prove(Expr(min) >= cast->type.min()) &&
```
Comment (psuriana) [inline src/Simplify.cpp]: Indentation: one tab too many

**Comment by jingpu** [comment]
@psuriana @dsharletg Thanks for the comments! I wonder if you have any insights on the Params checking assertions issue in my previous comment.

**Comment by abadams** [comment]
There's code in AddParameterChecks.cpp that seems to try to address the same thing by replacing use of the Param with code that includes explicit mins and maxes.... I'm confused now. Is this code not triggering in your case?

**Comment by jingpu** [comment]
@abadams You are right. The `param.constrained` variables added in AddParameterChecks.cpp should already solve the problem. I just need to include the let statements that define `param.constrained` when I want to simplify.

I can either make the PR just about "const_int_bounds() digging into Uint and Cast", or close the PR.

**Comment by abadams** [comment]
find_constant_bound is now much more aggressive in a way that I believe renders this issue moot

**Pull request closed without merging**